ethexe/network/src/db_sync/mod.rs

```diff
@@ -0,0 +1,931 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2024 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+mod ongoing;
+
+use crate::{
+    db_sync::ongoing::{
+        OngoingRequests, OngoingResponses, PeerFailed, PeerResponse, SendRequestError,
+        SendRequestErrorKind,
+    },
+    export::{Multiaddr, PeerId},
+    utils::ParityScaleCodec,
+};
+use ethexe_db::Database;
+use gear_core::ids::ProgramId;
+use gprimitives::H256;
+use libp2p::{
+    core::Endpoint,
+    request_response,
+    request_response::{InboundFailure, Message, OutboundFailure, ProtocolSupport},
+    swarm::{
+        CloseConnection, ConnectionDenied, ConnectionId, FromSwarm, NetworkBehaviour, THandler,
+        THandlerInEvent, THandlerOutEvent, ToSwarm,
+    },
+    StreamProtocol,
+};
+use parity_scale_codec::{Decode, Encode};
+use std::{
+    collections::{BTreeMap, BTreeSet, VecDeque},
+    task::{Context, Poll},
+    time::Duration,
+};
+
+const STREAM_PROTOCOL: StreamProtocol =
+    StreamProtocol::new(concat!("/ethexe/db-sync/", env!("CARGO_PKG_VERSION")));
```

Comment:
only two identical nodes could be synced?

Reply:
better to use constant version and bump when needed.

Reply:
Anyway, version management is for the future